### PR TITLE
issue/2756 Reduced suspend data usage

### PR DESCRIPTION
**Pull request by olivermartinfoster**
[#2756](https://github.com/adaptlearning/adapt_framework/issues/2756)
#### Changed
* Reduced suspend data usage using `Adapt.offlineStorage.serialize` and `Adapt.offlineStorage.deserialize`
* Bumped framework dependency to v5.4.*

Requires [#193](https://github.com/adaptlearning/adapt-contrib-spoor/pull/193) and [#2757](https://github.com/adaptlearning/adapt_framework/pull/2757)

### CODE REVIEW

File: js/adapt-assessmentArticleModel.js [hunk 1]
```diff
@@ -703,18 +701,10 @@ define([
...
-      for (id in indexByIdQuestions) {
-        if (indexByIdQuestions.hasOwnProperty(id)) {
-          var blockId = Adapt.findById(id).get('_parentId');
-          blockIds[blockId] = Adapt.findById(blockId);
-        }
-      }
-      var restoredChildrenModels = _.values(blockIds);
-
-      if (indexByIdQuestions) this.getChildren().models = restoredChildrenModels;
+      if (restoredChildrenModels && restoredChildrenModels.length) this.getChildren().models = restoredChildrenModels;
```
Comment (cahirodoherty-learningpool) [inline js/adapt-assessmentArticleModel.js:707]: Minor: Probably nicer to have this across multiple lines

**Comment by cahirodoherty-learningpool** [comment]
Minor comment: I'm assuming you will be running a lint on these files after this initial, functional PR. Just noticed some mixing of var/let/const.

File: js/adapt-assessmentArticleModel.js (hunk 1 above)
Comment (tomgreenfield) [reply js/adapt-assessmentArticleModel.js:707]: Does `restoredChildrenModels` even have to be checked as a truthy? `.map()` should always return an array.

**Comment by oliverfoster** [comment]
@cahirodoherty-learningpool 
I'm not intending to rewrite spoor to es6+ in this PR, just wanted this to be a feature addition. I can do that in a separate pr afterwards if that's alright?

File: js/adapt-assessmentArticleModel.js (hunk 1 above)
Comment (tomgreenfield) [reply js/adapt-assessmentArticleModel.js]: Notwithstanding the fact that this variable isn't even needed in the new code 😉.

File: js/adapt-assessmentArticleModel.js [hunk 2]
```diff
@@ -653,47 +654,44 @@ define([
...
-        }
+        blocks = state.questions.map(question => Adapt.findById(question._id).getParent());
       }
+      blocks = blocks.filter(block => {
+        const trackingId = block.get('_trackingId');
+        return Number.isInteger(trackingId) && trackingId >= 0;
+      });
+      const blockTrackingIds = blocks.map(block => block.get('_trackingId'));
+      const blockCompletion = blocks.map(block => {
+        let questions = block.findDescendantModels('question');
```
Comment (tomgreenfield) [inline js/adapt-assessmentArticleModel.js:674]: Nice – should probably update the JSDoc comments for `AdaptModel.findDescendantModels` at some point to reflect the fact you can pass in model types other than MPABC etc.

Comment (moloko) [reply js/adapt-assessmentArticleModel.js:674]: also could do with changing the jsdoc comments on that to remove the pluralisation of the sample values for the `descendants` parameter

**Review by lc-alexanderbenesch — APPROVED** [review]
👍

**Pull request merged**